query_optimizer/rules/PushDownSemiAntiJoin.cpp

```diff
@@ -0,0 +1,115 @@
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ **/
+
+#include "query_optimizer/rules/PushDownSemiAntiJoin.hpp"
+
+#include <vector>
+
+#include "query_optimizer/expressions/AttributeReference.hpp"
+#include "query_optimizer/expressions/ExpressionUtil.hpp"
+#include "query_optimizer/logical/HashJoin.hpp"
+#include "query_optimizer/logical/Logical.hpp"
+#include "query_optimizer/logical/PatternMatcher.hpp"
+
+namespace quickstep {
+namespace optimizer {
+
+namespace E = ::quickstep::optimizer::expressions;
+namespace L = ::quickstep::optimizer::logical;
+
+L::LogicalPtr PushDownSemiAntiJoin::applyToNode(const L::LogicalPtr &input) {
+  L::HashJoinPtr hash_join;
+
+  const std::vector<L::LogicalPtr> children = input->children();
+  std::vector<L::LogicalPtr> new_children;
+  bool has_changes = false;
+  for (const L::LogicalPtr& child : children) {
+    if (L::SomeHashJoin::MatchesWithConditionalCast(child, &hash_join)) {
+      if (hash_join->join_type() == L::HashJoin::kLeftSemiJoin
+          || hash_join->join_type() == L::HashJoin::kLeftAntiJoin) {
+        L::LogicalPtr new_child = pushDownSemiAntiJoin(hash_join);
+        if (new_child != child) {
+          has_changes = true;
+        }
+        new_children.push_back(new_child);
+      }
+    } else {
+      new_children.push_back(child);
+    }
+  }
+
+  if (has_changes) {
+    return input->copyWithNewChildren(new_children);
+  } else {
+    return input;
+  }
+}
+
+L::LogicalPtr PushDownSemiAntiJoin::pushDownSemiAntiJoin(
+    const L::HashJoinPtr &semi_anti_join) {
+  const L::LogicalPtr left_input = semi_anti_join->left();
+  std::vector<L::LogicalPtr> left_input_children = left_input->children();
+
+  if (!left_input_children.empty()) {
+    // Cannot push down a Filter down the right child of LeftOuterJoin.
+    std::vector<L::LogicalPtr>::size_type last_input_index = left_input_children.size();
+
+    std::vector<L::LogicalPtr>::size_type input_index = 0;
+    while (input_index < last_input_index) {
+      if (SubsetOfExpressions(semi_anti_join->left_join_attributes(),
+                              left_input_children[input_index]->getOutputAttributes())) {
+        break;
+      }
+      ++input_index;
+    }
+
+    if (input_index < last_input_index &&
+        semi_anti_join->residual_predicate() != nullptr) {
+      const std::vector<E::AttributeReferencePtr> input_child_output_attributes =
+          left_input_children[input_index]->getOutputAttributes();
+      const std::vector<E::AttributeReferencePtr> right_input_output_attributes =
```

Comment:
Could we mark both `input_child_output_attributes` and `right_input_output_attributes` as `reference`s, instead of doing deep copies? Thanks!